File: offer/JZ54 二叉搜索树的第k个节点/tree.java
import java.util.*;

/*
 * public class TreeNode {
 *   int val = 0;
 *   TreeNode left = null;
 *   TreeNode right = null;
 *   public TreeNode(int val) {
 *     this.val = val;
 *   }
 * }
 */

public class Solution {
    private ArrayList<Integer> preOrder = new ArrayList<>();
    private void preOrderTraverse(TreeNode root){
        if(root == null)
            return;
        preOrderTraverse(root.left);
        preOrder.add(root.val);
        preOrderTraverse(root.right);
    }
    public int KthNode (TreeNode root, int k) {
        preOrderTraverse(root);
        if(preOrder.size() < k || k <= 0)
            return -1;
        return preOrder.get(k - 1);
    }
}

File: 0144. Binary Tree Preorder Traversal/tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        if(root == nullptr)
            return {};
        vector<int> ret;
        stack<TreeNode*> stk;
        stk.push(root);
        while(!stk.empty()){
            TreeNode* top = stk.top();
            stk.pop();
            ret.push_back(top->val);
            if(top->right){
                stk.push(top->right);
            }
            if(top->left){
                stk.push(top->left);
            }
        }
        return ret;
    }
};